scipy/integrate/_tanhsinh.py

```diff
@@ -934,31 +944,42 @@ def _nsum(f, a, b, step=1, args=(), log=False, maxterms=int(2**20), atol=None,
 
             f(x: ndarray, *args) -> ndarray
 
-         where each element of ``x`` is a finite real and ``args`` is a tuple,
-         which may contain an arbitrary number of arrays that are broadcastable
-         with `x`. `f` must represent a smooth, positive, and monotone decreasing
-         function of `x`; `_nsum` performs no checks to verify that these conditions
-         are met and may return erroneous results if they are violated.
-    a, b : array_like
+        where each element of ``x`` is a finite real and ``args`` is a tuple,
+        which may contain an arbitrary number of arrays that are broadcastable
+        with ``x``.
+         
+        `f` must be an elementwise function: each element ``f(x)[i]``
+        must equal ``f(x[i])`` for all indices ``i``. It must not mutate the
+        array ``x`` or the arrays in ``args``, and it must return NaN where
+        the argument is NaN.
+        
+        `f` must represent a smooth, positive, and monotone decreasing
+        function of `x` defined at *all reals* between `a` and `b`.          
+        `nsum` performs no checks to verify that these conditions
+        are met and may return erroneous results if they are violated.
+    a, b : real number array
```

Comment:
The function accepts lists because SciPy has not wholely switched to `SCIPY_ARRAY_API=1` being the default and only behavior. If and when that change is made, this function would transition with all other functions in SciPy.

I didn't document it as accepting `array_like`s because that is only slightly more specific than stating that it accepts `object`s, (according to either the NumPy glossary or `numpy.typing` definition). 

"real array_like" might actually be specific enough, suggesting that it accepts any object that can be converted into an array of real numbers, but I just hesitate to use the term `array_like` for the reason above.

There are a few functions in SciPy that err in the other direction, suggesting that they accept any `array_like` but actually failing except when specific conditions are met.